src/util/chunk_fs.js

```diff
@@ -59,7 +59,8 @@ class ChunkFS extends stream.Transform {
     }
 
     async _flush(callback) {
-        this._flush_buffers(callback);
+        // wait before the last writev to finish
+        await this._flush_buffers(callback);
```

Comment:
Hey @utkarsh-pro 
_flush() is not really an async function in nodejs streams, and the caller will not await for its promise, so it's a bit misleading. It passes a callback function instead - which we handle correctly inside _flush_buffers. The fact that we defined _flush() itself to be async was accidental. So if you get complaints on unhandled rejection of the promise returned by _flush_buffers() I would instead add a noop catch handler to it like `_flush_buffers(callback).catch(noop)` - you can see how we do it in other places - 

https://github.com/noobaa/noobaa-core/blob/cfbf726191039f6da8e832f680851e139a0ca4ce/src/server/system_services/system_store.js#L374
https://github.com/noobaa/noobaa-core/blob/7a31c5d46a96c4acce8b4dc34406dffaa572ed69/src/rpc/ice.js#L1323
https://github.com/noobaa/noobaa-core/blob/db1bc28d741943652e8d17bc50afe3bc74a7416d/src/util/postgres_client.js#L1466

Reply:
@guymguym, actually that's true but for recent node versions. In node v16, although undocumented, it does calls `then` if a promise is returned. https://github.com/nodejs/node/blob/418ff70b810f0e7112d48baaa72932a56cfa213b/lib/internal/streams/transform.js#L131

Here's my understanding code, please correct me if this is wrong:
1. We have async `_flush` (accidental probably). This function always returns a promise.
2. Node sees that the returned value has `.then` attached to it.
3. As soon as the promise is resolved (which without `await` on our `flush_buffer` resolves almost immediately).
4. This leads to call `this.push(null)` in the node source.
5. Call to `this.push(null)` leads to emitting of `readable` event.
6. That leads to calling `next` which realises that oops, we have read everything.
7. Emits `end` event.
8. `finished()` already encountered the `finish` event by this time, only thing holding it back from getting resolved was the `end` event from the transform stream.
9. Above emit of `end` event caused the `finished()` to resolve.

And that's how I think we exit the `finished` before we intended. As mentioned, this causes the test to non-deterministic, sometimes promise resolves first (when read buffer indeed requires flushing) and sometimes `callback` is invoked first (desired).

Reply:
Here is a sample which won't work in node v16.14.0 (but would in recent in v14):

```javascript
const util = require("util");
const timeout = util.promisify(setTimeout);
const crypto = require("crypto");
const { Readable, Transform, pipeline, finished } = require("stream");
const pipelineAsync = util.promisify(pipeline);
const finishedAsync = util.promisify(finished);

async function P(...streams) {
  return pipelineAsync(...streams).then(() => {
    streams[streams.length - 1].resume();
  });
}

class TargetHash {
  constructor() {
    this.hash = crypto.createHash("md5");
  }

  digest() {
    return this.hash.digest("hex");
  }

  async writev(_config, buffers) {
    await timeout(500);
    for (const buf of buffers) this.hash.update(buf);
  }
}

class T extends Transform {
  constructor({ targetHash }) {
    super();
    this.th = targetHash;
  }

  async _transform(chunk, encoding, callback) {
    await this.th.writev({}, [chunk]);
    callback();
  }

  async _flush(callback) {
    console.log("flush start");
    this._flush_work(callback);
    console.log("flush end");
  }

  async _flush_work(cb) {
    console.log("flush_work start");
    await this.th.writev({}, [Buffer.from("EOF")]);
    console.log("flush_work end");

    console.log("flush_work cb start");
    const res = cb();
    console.log("flush_work cb end");
    return res;
  }
}

async function main() {
  const th = new TargetHash();
  const t = new T(
    { targetHash: th },
  );
  const r = Readable.from(
    (async function* () {
      for (let i = 0; i < 1e1; i++) {
        yield i.toString();
      }
    })()
  );

  console.log("pipeline start");
  await P(r, t);
  await finishedAsync(t);
  console.log("DIGEST:", th.digest());
  console.log("pipeline finished");
}

main();

```